### PR TITLE
Improve parse error reporting

### PR DESCRIPTION
**Pull request by Lupus**
Better location + error context

### CODE REVIEW

**Review by c-cube — CHANGES_REQUESTED** [review]
That looks quite nice! Two things:
- little expect test to see the failure?
- can we replace `Parsing_error of string * string * Loc.t` with an inline record? `Parsing_error of { msg: string; context: string; loc: Loc.t }`? Helps readability :)

**Comment by c-cube** [comment]
You are on a roll, that's awesome :)

File: src/compilerlib/pb_exception.ml [hunk 1]
```diff
@@ -230,14 +234,16 @@ let missing_field_label ~field_name ~message_name =
...
 
-let protoc_parsing_error e loc =
-  raise (Compilation_error (Parsing_error (string_of_error e, loc)))
+let protoc_parsing_error e loc context =
+  raise
+    (Compilation_error (Parsing_error { msg = string_of_error e; context; loc }))
 
-let unknown_parsing_error detail loc =
-  raise (Compilation_error (Parsing_error (detail, loc)))
+let unknown_parsing_error ~msg ~context loc =
```
Comment (c-cube) [inline src/compilerlib/pb_exception.ml:245]: :chef kiss:

**Comment by c-cube** [comment]
Ah you might need to depend on `result` on 4.03? 

If it's too much of a chore I think this could also be a reasonable time to move the lower bound to 4.08. It's old enough at this point.

**Comment by Lupus** [comment]
I'm voting for 4.08 bump 😊

**Comment by c-cube** [comment]
Thank you! :)

**Pull request merged**